File: src/main/java/org/oewntk/wndb/in/Factory.java
/*
 * Copyright (c) 2021-2021. Bernard Bou.
 */

package org.oewntk.wndb.in;

import kotlin.Pair;
import org.oewntk.model.*;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;
import java.util.Collection;
import java.util.Map.Entry;
import java.util.function.Supplier;

/**
 * Model factory
 */
public class Factory implements Supplier<Model>
{
    private final File inDir;

    private final File inDir2;

    /**
     * Constructor
     *
     * @param inDir  WNDB dir
     * @param inDir2 extra WNDB dir
     */
    public Factory(final File inDir, final File inDir2)
    {
        this.inDir = inDir;
        this.inDir2 = inDir2;
    }

    @Override
    public Model get()
    {
        CoreModel coreModel = new CoreFactory(inDir).get();
        if (coreModel == null)
        {
            return null;
        }

        try
        {
            // verb frames and templates
            Collection<VerbFrame> verbFramesById = new VerbFrameParser(inDir2).parse();
            Collection<VerbTemplate> verbTemplatesById = new VerbTemplateParser(inDir2).parse();
            Collection<Pair<String, Integer[]>> senseToVerbTemplates = new SenseToVerbTemplatesParser(inDir).parse();

            // tag counts
            Collection<Pair<String, TagCount>> senseToTagCounts = new SenseToTagCountsParser(inDir).parse();

            return new Model(coreModel, verbFramesById, verbTemplatesById, senseToVerbTemplates, senseToTagCounts).setSources(inDir, inDir2);
        }
        catch (IOException e)
        {
            e.printStackTrace(Tracing.psErr);
            return null;
        }
    }

    /**
     * Make model
     *
     * @param inDir  WNDB dir
     * @param inDir2 extra WNDB dir
     * @return model
     */
    static public Model makeModel(File inDir, File inDir2)
    {
        return new Factory(inDir, inDir2).get();
    }

    /**
     * Make model
     *
     * @param dirPath1 WNDB dir path
     * @param dirPath2 extra WNDB dir path
     * @return core model
     */
    static public Model makeModel(final String dirPath1, final String dirPath2)
    {
        File inDir = new File(dirPath1);
        File inDir2 = new File(dirPath2);
        return new Factory(inDir, inDir2).get();
    }

    /**
     * Make model
     *
     * @param args command-line arguments
     * @return core model
     */
    static public Model makeModel(String[] args)
    {
        File inDir = new File(args[0]);
        File inDir2 = new File(args[1]);
        return makeModel(inDir, inDir2);
    }

    /**
     * Main
     *
     * @param args command-line arguments
     */
    static public void main(String[] args)
    {
        String dirPath2 = args[args.length - 1]; // last
        for (int i = 0; i < args.length - 1; i++)
        {
            Model model = makeModel(args[i], dirPath2);
            Tracing.psInfo.printf("[Model] %s%n%s%n%s%n", Arrays.toString(model.getSources()), model.info(), model.counts());
        }
    }
}

File: src/main/java/org/oewntk/wndb/in/Factory.kt
/*
 * Copyright (c) 2021-2024. Bernard Bou.
 */
package org.oewntk.wndb.`in`

import org.oewntk.model.Injector
import org.oewntk.model.Model
import org.oewntk.model.ModelInfo
import org.oewntk.model.TagCount
import java.io.File
import java.io.IOException
import java.util.function.Supplier

/**
 * Model factory
 *
 * @property inDir  WNDB dir
 * @property inDir2 extra WNDB dir
 */
class Factory(
    private val inDir: File,
    private val inDir2: File?,
    private val inverses: Boolean = false,
    private val verbose: Boolean = false,
) : Supplier<Model?> {

    override fun get(): Model? {
        val coreModel = CoreFactory(inDir, inverses = inverses, verbose = verbose).get() ?: return null

        try {
            // verb frames and templates
            val verbFramesById = VerbFrameParser(inDir2 ?: inDir).parse() // orig:verbFrames.txt gen:verbFrames.txt
            val verbTemplatesById = VerbTemplateParser(inDir2 ?: inDir).parse() // templates.txt

            // sense to verb templates
            val senseToVerbTemplates = SenseToVerbTemplatesParser(inDir).parse() // sentidx.vrb

            // tag counts
            val senseToTagCounts: Collection<Pair<String, TagCount>> = SenseToTagCountsParser(inDir).parse() // cntlist.rev

            return Model(coreModel, verbFramesById, verbTemplatesById, Injector(senseToVerbTemplates, senseToTagCounts))
                .apply {
                    source = inDir.absolutePath
                    source2 = inDir2?.absolutePath
                }

        } catch (e: IOException) {
            e.printStackTrace(Tracing.psErr)
            return null
        }
    }

    companion object {

        /**
         * Make model
         *
         * @param inDir  WNDB dir
         * @param inDir2 extra WNDB dir
         * @return model
         */
        private fun makeModel(inDir: File, inDir2: File?, inverses: Boolean = false, verbose: Boolean = false): Model? {
            return Factory(inDir, inDir2, inverses = inverses, verbose = verbose).get()
        }

        /**
         * Make model
         *
         * @param args command-line arguments
         * @return core model
         */
        fun makeModel(args: Array<String>): Model? {
            var iArg = 0
            var verbose = false
            var inverses = false
            if (args[iArg] == "--verbose") {
                verbose = true
                iArg++
            }
            if ("--inverses" == args[iArg]) {
                inverses = true
                iArg++
            }
            val inDir = File(args[iArg])
            iArg++
            val inDir2 = if (iArg < args.size) File(args[iArg]) else null
            return makeModel(inDir, inDir2, inverses = inverses, verbose = verbose)
        }

        /**
         * Main
         *
         * @param args command-line arguments
         */
        @JvmStatic
        fun main(args: Array<String>) {
            val model = makeModel(args)
            Tracing.psInfo.printf("[Model] %s%n%s%n%s%n", model!!, model.info(), ModelInfo.counts(model))
        }
    }
}